wasm/datatypes/stack.py

```diff
@@ -0,0 +1,86 @@
+from typing import (
+    Generic,
+    Iterable,
+    List,
+    NamedTuple,
+    Tuple,
+    TypeVar,
+)
+import uuid
+
+from wasm.typing import (
+    TValue,
+)
+
+from .indices import (
+    LabelIdx,
+)
+from .instructions import (
+    InstructionSequence,
+)
+from .module import (
+    ModuleInstance,
+)
+
+TStackItem = TypeVar('TStackItem')
+
+
+class BaseStack(Generic[TStackItem]):
+    _stack: List[TStackItem]
+
+    def __init__(self) -> None:
+        self._stack = []
+
+    def __len__(self) -> int:
+        return len(self._stack)
+
+    def __bool__(self) -> bool:
+        return bool(self._stack)
+
+    def __iter__(self) -> Iterable[TStackItem]:
+        return iter(self._stack)
+
+    def pop(self) -> TStackItem:
+        return self._stack.pop()
+
+    def pop2(self) -> Tuple[TStackItem, TStackItem]:
+        return self._stack.pop(), self._stack.pop()
+
+    def pop3(self) -> Tuple[TStackItem, TStackItem, TStackItem]:
+        return self._stack.pop(), self._stack.pop(), self._stack.pop()
+
+    def push(self, value: TStackItem) -> None:
+        self._stack.append(value)
+
+    def peek(self) -> TStackItem:
+        return self._stack[-1]
+
+
+class ValueStack(BaseStack[TValue]):
+    pass
+
+
+class Frame(NamedTuple):
+    id: uuid.UUID
+    module: ModuleInstance
+    locals: List[TValue]
+    instructions: InstructionSequence
+    arity: int
+    height: int
+
+
+class FrameStack(BaseStack[Frame]):
+    pass
+
+
+class Label(NamedTuple):
+    arity: int
+    height: int
+    instructions: InstructionSequence
+    is_loop: bool
+    frame_id: uuid.UUID
```

Comment:
This is one of the things that isn't in the spec but that I needed to implement for the frame/stack juggling to work correctly.  Basically, when a new `Frame` lands on the frame stack, the label stack should be cleared.  However, once the new frame exits and the old frame resumes, the old label stack needs to be restored.

I accomplished this by tieing the labels to the frame their were created in allowing us to treat the label stack as empty if the top label's frame isn't the top frame.